File: mobile/lib/data/selectors.dart
import 'mock_data.dart';
import '../models/models.dart';

int sumI(Iterable<int> xs) => xs.fold(0, (a, b) => a + b);

class GroupSummary {
  final int totalMembers, activeMembers, newMembers, totalShares, totalSavings, disbursed,
      outstanding, repayments, revenue, expenses, netProfit, projectCapital, insuranceContributions;
  final double par;
  final int activeProjects, pendingClaims, membersCovered;
  GroupSummary({
    required this.totalMembers, required this.activeMembers, required this.newMembers,
    required this.totalShares, required this.totalSavings, required this.disbursed,
    required this.outstanding, required this.repayments, required this.revenue, required this.expenses,
    required this.netProfit, required this.projectCapital, required this.insuranceContributions,
    required this.par, required this.activeProjects, required this.pendingClaims, required this.membersCovered,
  });
}

GroupSummary groupSummary() {
  final totalShares = sumI(mock.shares.map((s) => s.totalValue));
  final totalSavings = sumI(mock.savingsAccounts.map((a) => a.balance));
  final disbursed = sumI(mock.loans.where((l) => l.disbursementDate != null).map((l) => l.principal));
  final outstanding = sumI(mock.loans.map((l) => l.outstanding));
  final repayments = sumI(mock.repayments.map((r) => r.totalPaid));
  int acc(String code) => mock.accounts.firstWhere((a) => a.code == code, orElse: () => Account(id: '', code: '', name: '', type: '', balance: 0)).balance;
  final revenue = acc('4000') + acc('4100') + acc('4200');
  final expenses = sumI(mock.accounts.where((a) => a.type == 'expense').map((a) => a.balance));
  final parLoans = mock.loans.where((l) => l.status == 'overdue' || l.status == 'defaulted');
  final par = outstanding == 0 ? 0.0 : sumI(parLoans.map((l) => l.outstanding)) / outstanding;
  return GroupSummary(
    totalMembers: mock.members.length,
    activeMembers: mock.activeMembers.length,
    newMembers: mock.members.where((m) => DateTime.now().difference(DateTime.parse(m.registrationDate)).inDays < 60).length,
    totalShares: totalShares, totalSavings: totalSavings, disbursed: disbursed, outstanding: outstanding,
    repayments: repayments, revenue: revenue, expenses: expenses, netProfit: revenue - expenses,
    projectCapital: sumI(mock.projects.map((p) => p.capitalRaised)),
    insuranceContributions: sumI(mock.insuranceContributions.map((c) => c.amount)),
    par: par,
    activeProjects: mock.projects.where((p) => p.status == 'active').length,
    pendingClaims: mock.insuranceClaims.where((c) => c.status == 'submitted' || c.status == 'under_review').length,
    membersCovered: mock.insuranceAccounts.where((a) => a.status == 'active').length,
  );
}

List<Map<String, dynamic>> cashFlowSeries() {
  const months = ['Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep'];
  return List.generate(months.length, (i) => {
        'month': months[i],
        'inflow': 4200000 + i * 380000 + (i.isOdd ? 600000 : 0),
        'outflow': 3100000 + i * 240000 + (i % 3 != 0 ? 300000 : 0),
      });
}

List<Map<String, dynamic>> savingsVsLoansSeries() {
  const months = ['Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep'];
  return List.generate(months.length, (i) => {
        'month': months[i],
        'savings': 52000000 + i * 3400000,
        'loans': 40000000 + i * 3900000,
      });
}

List<Map<String, dynamic>> contributionTrend() {
  const months = ['Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep'];
  return List.generate(months.length, (i) => {'month': months[i], 'amount': 620000 + i * 45000 + (i.isOdd ? 40000 : 0)});
}

Map<String, int> loanStatusBreakdown() {
  final map = <String, int>{};
  for (final l in mock.loans) {
    map[l.status] = (map[l.status] ?? 0) + 1;
  }
  return map;
}

class MemberPosition {
  final int shareValue, shareQty, savingsBalance, loanOutstanding, projectInvestment, projectReturn, profit;
  final String? savingsAccountId;
  final Loan? activeLoan;
  final InsuranceAccount? insurance;
  final List<Loan> loans;
  final List<ProjectInvestment> investments;
  MemberPosition({
    required this.shareValue, required this.shareQty, required this.savingsBalance,
    required this.loanOutstanding, required this.projectInvestment, required this.projectReturn,
    required this.profit, this.savingsAccountId, this.activeLoan, this.insurance,
    required this.loans, required this.investments,
  });
}

MemberPosition memberPosition(String memberId) {
  final ms = mock.shares.where((s) => s.memberId == memberId).toList();
  final sa = mock.savingsAccounts.where((a) => a.memberId == memberId).toList();
  final ml = mock.loans.where((l) => l.memberId == memberId).toList();
  final invest = mock.projectInvestments.where((p) => p.memberId == memberId).toList();
  final ins = mock.insuranceAccounts.where((a) => a.memberId == memberId).toList();
  final shareValue = sumI(ms.map((s) => s.totalValue));
  Loan? active;
  for (final l in ml) {
    if (['active', 'overdue', 'disbursed'].contains(l.status)) {
      active = l;
      break;
    }
  }
  return MemberPosition(
    shareValue: shareValue,
    shareQty: sumI(ms.map((s) => s.quantity)),
    savingsBalance: sa.isNotEmpty ? sa.first.balance : 0,
    savingsAccountId: sa.isNotEmpty ? sa.first.id : null,
    activeLoan: active,
    loanOutstanding: sumI(ml.map((l) => l.outstanding)),
    projectInvestment: sumI(invest.map((p) => p.amount)),
    projectReturn: sumI(invest.map((p) => p.amount + p.profitShare)),
    profit: sumI(invest.map((p) => p.profitShare)) + (shareValue / 1000000 * 42000).round(),
    insurance: ins.isNotEmpty ? ins.first : null,
    loans: ml,
    investments: invest,
  );
}

List<Txn> memberTransactions(String memberId) => mock.transactions.where((t) => t.memberId == memberId).toList();

File: src/mock/selectors.ts
import {
  accounts, activeMembers, insuranceAccounts, insuranceClaims, insuranceContributions,
  loanRepayments, loans, members, projectInvestments, projects, savingsAccounts, shares,
  transactions,
} from './data'

export const sum = (arr: number[]) => arr.reduce((a, b) => a + b, 0)

export function groupSummary() {
  const totalShares = sum(shares.map((s) => s.totalValue))
  const totalSavings = sum(savingsAccounts.map((a) => a.balance))
  const disbursed = sum(loans.filter((l) => l.disbursementDate).map((l) => l.principal))
  const outstanding = sum(loans.map((l) => l.outstanding))
  const repayments = sum(loanRepayments.map((r) => r.totalPaid))
  const interestIncome = accounts.find((a) => a.code === '4000')?.balance ?? 0
  const projectIncome = accounts.find((a) => a.code === '4100')?.balance ?? 0
  const feesIncome = accounts.find((a) => a.code === '4200')?.balance ?? 0
  const expenses = sum(accounts.filter((a) => a.type === 'expense').map((a) => a.balance))
  const revenue = interestIncome + projectIncome + feesIncome
  const netProfit = revenue - expenses
  const parLoans = loans.filter((l) => l.status === 'overdue' || l.status === 'defaulted')
  const par = outstanding ? sum(parLoans.map((l) => l.outstanding)) / outstanding : 0

  return {
    totalMembers: members.length,
    activeMembers: activeMembers.length,
    newMembers: members.filter((m) => Date.now() - +new Date(m.registrationDate) < 60 * 864e5).length,
    totalShares,
    totalSavings,
    disbursed,
    outstanding,
    repayments,
    revenue,
    expenses,
    netProfit,
    par,
    activeProjects: projects.filter((p) => p.status === 'active').length,
    projectCapital: sum(projects.map((p) => p.capitalRaised)),
    insuranceContributions: sum(insuranceContributions.map((c) => c.amount)),
    pendingClaims: insuranceClaims.filter((c) => ['submitted', 'under_review'].includes(c.status)).length,
    membersCovered: insuranceAccounts.filter((a) => a.status === 'active').length,
  }
}

export function cashFlowSeries() {
  const months = ['Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep']
  return months.map((m, i) => ({
    month: m,
    inflow: 4_200_000 + i * 380_000 + (i % 2 ? 600_000 : 0),
    outflow: 3_100_000 + i * 240_000 + (i % 3 ? 300_000 : 0),
  }))
}

export function savingsVsLoansSeries() {
  const months = ['Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep']
  return months.map((m, i) => ({
    month: m,
    savings: 52_000_000 + i * 3_400_000,
    loans: 40_000_000 + i * 3_900_000,
  }))
}

export function loanStatusBreakdown() {
  const map = new Map<string, number>()
  loans.forEach((l) => map.set(l.status, (map.get(l.status) ?? 0) + 1))
  return [...map.entries()].map(([status, count]) => ({ status, count }))
}

export function contributionTrend() {
  const months = ['Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep']
  return months.map((m, i) => ({ month: m, amount: 620_000 + i * 45_000 + (i % 2 ? 40_000 : 0) }))
}

/* -------- member-scoped -------- */
export function memberPosition(memberId: string) {
  const memberShares = shares.filter((s) => s.memberId === memberId)
  const savingsAcc = savingsAccounts.find((a) => a.memberId === memberId)
  const memberLoans = loans.filter((l) => l.memberId === memberId)
  const activeLoan = memberLoans.find((l) => ['active', 'overdue', 'disbursed'].includes(l.status))
  const invest = projectInvestments.filter((p) => p.memberId === memberId)
  const insurance = insuranceAccounts.find((a) => a.memberId === memberId)

  return {
    shareValue: sum(memberShares.map((s) => s.totalValue)),
    shareQty: sum(memberShares.map((s) => s.quantity)),
    savingsBalance: savingsAcc?.balance ?? 0,
    savingsAccountId: savingsAcc?.id,
    activeLoan,
    loanOutstanding: sum(memberLoans.map((l) => l.outstanding)),
    projectInvestment: sum(invest.map((p) => p.amount)),
    projectReturn: sum(invest.map((p) => p.amount + p.profitShare)),
    profit: sum(invest.map((p) => p.profitShare)) + Math.round((sum(memberShares.map((s) => s.totalValue)) / 1_000_000) * 42_000),
    insurance,
    loans: memberLoans,
    investments: invest,
  }
}

export function memberTransactions(memberId: string) {
  return transactions.filter((t) => t.memberId === memberId)
}
